{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running test3-baseline over chr9:5000000-7000000 opts \n",
      "\n",
      "            ../../gatk HaplotypeCaller -R ../ref/Homo_sapiens_assembly38.fasta -I data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam -intervals chr9:5000000-7000000             --likelihood-calculation-engine FlowBased             --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false             -O output/test3/baseline/test3-baseline.vcf             --bam-output output/test3/baseline/test3-baseline.bam                  \n",
      "21:15:43.511 INFO  ProgressMeter - Starting traversal\n",
      "21:15:43.511 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "21:15:53.515 INFO  ProgressMeter -         chr9:5118303              0.2                   740           4438.7\n",
      "21:16:03.523 INFO  ProgressMeter -         chr9:5282712              0.3                  1850           5546.7\n",
      "21:16:13.524 INFO  ProgressMeter -         chr9:5419271              0.5                  2820           5637.6\n",
      "21:16:23.673 INFO  ProgressMeter -         chr9:5539189              0.7                  3710           5542.6\n",
      "21:16:33.746 INFO  ProgressMeter -         chr9:5669410              0.8                  4680           5589.7\n",
      "21:16:43.786 INFO  ProgressMeter -         chr9:5777213              1.0                  5500           5474.9\n",
      "21:16:53.848 INFO  ProgressMeter -         chr9:5871568              1.2                  6230           5314.4\n",
      "21:17:03.911 INFO  ProgressMeter -         chr9:5990969              1.3                  7130           5320.9\n",
      "21:17:14.086 INFO  ProgressMeter -         chr9:6093039              1.5                  7920           5246.5\n",
      "21:17:24.131 INFO  ProgressMeter -         chr9:6215727              1.7                  8850           5277.3\n",
      "21:17:34.147 INFO  ProgressMeter -         chr9:6338809              1.8                  9780           5303.9\n",
      "21:17:44.150 INFO  ProgressMeter -         chr9:6437727              2.0                 10540           5242.1\n",
      "21:17:54.231 INFO  ProgressMeter -         chr9:6553017              2.2                 11440           5250.9\n",
      "21:18:04.333 INFO  ProgressMeter -         chr9:6646411              2.3                 12190           5193.8\n",
      "21:18:14.353 INFO  ProgressMeter -         chr9:6734099              2.5                 12910           5135.2\n",
      "21:18:24.396 INFO  ProgressMeter -         chr9:6843876              2.7                 13780           5139.1\n",
      "21:18:34.496 INFO  ProgressMeter -         chr9:6940946              2.8                 14480           5081.1\n",
      "21:18:42.047 INFO  ProgressMeter -         chr9:6999635              3.0                 14963           5028.6\n",
      "21:18:42.048 INFO  ProgressMeter - Traversal complete. Processed 14963 total regions in 3.0 minutes.\n",
      "{'_intervals': 'chr9:5000000-7000000',\n",
      " '_opts': '',\n",
      " '_test': 'test3',\n",
      " '_tver': 'baseline',\n",
      " '_vcf': 'output/test3/baseline/test3-baseline.vcf',\n",
      " 'indels': '833',\n",
      " 'mnps': '0',\n",
      " 'multiallelic-sites': '14',\n",
      " 'multiallelic-snp-sites': '1',\n",
      " 'no-alts': '0',\n",
      " 'others': '0',\n",
      " 'records': '3848',\n",
      " 'samples': '1',\n",
      " 'snps': '3015'}\n",
      "Running test3-collapse12m over chr9:5000000-7000000 opts --ultima-assembly-collapse-hmer-size 12 --ultima-flow-matrix-mods 10,12,11,12\n",
      "\n",
      "            ../../gatk HaplotypeCaller -R ../ref/Homo_sapiens_assembly38.fasta -I data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam -intervals chr9:5000000-7000000             --likelihood-calculation-engine FlowBased             --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false             -O output/test3/collapse12m/test3-collapse12m.vcf             --bam-output output/test3/collapse12m/test3-collapse12m.bam             --ultima-assembly-collapse-hmer-size 12 --ultima-flow-matrix-mods 10,12,11,12     \n",
      "21:18:49.517 INFO  ProgressMeter - Starting traversal\n",
      "21:18:49.517 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "21:18:59.892 INFO  ProgressMeter -         chr9:5127017              0.2                   800           4626.5\n",
      "21:19:09.950 INFO  ProgressMeter -         chr9:5293421              0.3                  1920           5637.9\n",
      "21:19:20.072 INFO  ProgressMeter -         chr9:5430699              0.5                  2910           5714.3\n",
      "21:19:30.099 INFO  ProgressMeter -         chr9:5551118              0.7                  3800           5618.4\n",
      "21:19:40.133 INFO  ProgressMeter -         chr9:5684721              0.8                  4810           5701.8\n",
      "21:19:50.210 INFO  ProgressMeter -         chr9:5786456              1.0                  5580           5516.3\n",
      "21:20:00.255 INFO  ProgressMeter -         chr9:5879714              1.2                  6290           5335.2\n",
      "21:20:10.820 INFO  ProgressMeter -         chr9:6009681              1.4                  7270           5365.1\n",
      "21:20:20.893 INFO  ProgressMeter -         chr9:6112299              1.5                  8080           5305.6\n",
      "21:20:30.952 INFO  ProgressMeter -         chr9:6239290              1.7                  9010           5329.5\n",
      "21:20:40.961 INFO  ProgressMeter -         chr9:6368532              1.9                 10010           5389.3\n",
      "21:20:50.972 INFO  ProgressMeter -         chr9:6460040              2.0                 10710           5290.8\n",
      "21:21:01.014 INFO  ProgressMeter -         chr9:6578474              2.2                 11650           5315.7\n",
      "21:21:11.058 INFO  ProgressMeter -         chr9:6677405              2.4                 12450           5277.6\n",
      "21:21:21.301 INFO  ProgressMeter -         chr9:6780143              2.5                 13290           5253.5\n",
      "21:21:31.303 INFO  ProgressMeter -         chr9:6914860              2.7                 14270           5292.2\n",
      "21:21:41.398 INFO  ProgressMeter -         chr9:6989204              2.9                 14870           5190.8\n",
      "21:21:43.875 INFO  ProgressMeter -         chr9:6999635              2.9                 14963           5149.1\n",
      "21:21:43.875 INFO  ProgressMeter - Traversal complete. Processed 14963 total regions in 2.9 minutes.\n",
      "{'_intervals': 'chr9:5000000-7000000',\n",
      " '_opts': '--ultima-assembly-collapse-hmer-size 12 --ultima-flow-matrix-mods '\n",
      "          '10,12,11,12',\n",
      " '_test': 'test3',\n",
      " '_tver': 'collapse12m',\n",
      " '_vcf': 'output/test3/collapse12m/test3-collapse12m.vcf',\n",
      " 'indels': '777',\n",
      " 'mnps': '0',\n",
      " 'multiallelic-sites': '13',\n",
      " 'multiallelic-snp-sites': '1',\n",
      " 'no-alts': '0',\n",
      " 'others': '0',\n",
      " 'records': '3765',\n",
      " 'samples': '1',\n",
      " 'snps': '2988'}\n",
      "Running test3-collapse10m over chr9:5000000-7000000 opts --ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods 11,10,12,10\n",
      "\n",
      "            ../../gatk HaplotypeCaller -R ../ref/Homo_sapiens_assembly38.fasta -I data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam -intervals chr9:5000000-7000000             --likelihood-calculation-engine FlowBased             --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false             -O output/test3/collapse10m/test3-collapse10m.vcf             --bam-output output/test3/collapse10m/test3-collapse10m.bam             --ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods 11,10,12,10     \n",
      "21:21:51.166 INFO  ProgressMeter - Starting traversal\n",
      "21:21:51.166 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "21:22:01.226 INFO  ProgressMeter -         chr9:5118303              0.2                   740           4413.5\n",
      "21:22:11.237 INFO  ProgressMeter -         chr9:5281662              0.3                  1840           5500.5\n",
      "21:22:21.238 INFO  ProgressMeter -         chr9:5414773              0.5                  2790           5566.6\n",
      "21:22:31.362 INFO  ProgressMeter -         chr9:5537029              0.7                  3690           5508.0\n",
      "21:22:41.482 INFO  ProgressMeter -         chr9:5665940              0.8                  4650           5545.0\n",
      "21:22:51.541 INFO  ProgressMeter -         chr9:5777213              1.0                  5500           5465.8\n",
      "21:23:02.135 INFO  ProgressMeter -         chr9:5872912              1.2                  6240           5275.5\n",
      "21:23:12.352 INFO  ProgressMeter -         chr9:5993110              1.4                  7150           5284.2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21:23:22.617 INFO  ProgressMeter -         chr9:6096894              1.5                  7960           5222.5\n",
      "21:23:32.685 INFO  ProgressMeter -         chr9:6226129              1.7                  8920           5271.9\n",
      "21:23:42.782 INFO  ProgressMeter -         chr9:6354140              1.9                  9900           5321.8\n",
      "21:23:53.077 INFO  ProgressMeter -         chr9:6457515              2.0                 10690           5261.2\n",
      "21:24:03.141 INFO  ProgressMeter -         chr9:6575448              2.2                 11630           5287.4\n",
      "21:24:13.238 INFO  ProgressMeter -         chr9:6673663              2.4                 12420           5245.2\n",
      "21:24:23.278 INFO  ProgressMeter -         chr9:6770663              2.5                 13210           5210.6\n",
      "21:24:33.300 INFO  ProgressMeter -         chr9:6894385              2.7                 14130           5229.0\n",
      "21:24:43.458 INFO  ProgressMeter -         chr9:6970904              2.9                 14720           5126.2\n",
      "21:24:47.425 INFO  ProgressMeter -         chr9:6999635              2.9                 14963           5093.5\n",
      "21:24:47.425 INFO  ProgressMeter - Traversal complete. Processed 14963 total regions in 2.9 minutes.\n",
      "{'_intervals': 'chr9:5000000-7000000',\n",
      " '_opts': '--ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods '\n",
      "          '11,10,12,10',\n",
      " '_test': 'test3',\n",
      " '_tver': 'collapse10m',\n",
      " '_vcf': 'output/test3/collapse10m/test3-collapse10m.vcf',\n",
      " 'indels': '787',\n",
      " 'mnps': '0',\n",
      " 'multiallelic-sites': '12',\n",
      " 'multiallelic-snp-sites': '1',\n",
      " 'no-alts': '0',\n",
      " 'others': '0',\n",
      " 'records': '3762',\n",
      " 'samples': '1',\n",
      " 'snps': '2975'}\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import io\n",
    "import pprint\n",
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "# operations\n",
    "OP_ANAL=True\n",
    "OP_PRINT_CMD=True\n",
    "OP_TVER_ONLY=\"\"\n",
    "\n",
    "# globals\n",
    "GATK=\"../../gatk\"\n",
    "OUT=\"output\"\n",
    "REF=\"../ref/Homo_sapiens_assembly38.fasta\"\n",
    "\n",
    "#READS=\"data/chr20.flow.bam\"\n",
    "#INTERVALS=\"chr20:6000000-7000000\"\n",
    "#TESTNAME=\"test1\"\n",
    "\n",
    "#READS=\"data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam\"\n",
    "#INTERVALS=\"chr9:1700000-2000000\"\n",
    "#TESTNAME=\"test2\"\n",
    "\n",
    "READS=\"data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam\"\n",
    "INTERVALS=\"chr9:5000000-7000000\"\n",
    "TESTNAME=\"test3\"\n",
    "\n",
    "# functions\n",
    "def HaplotypeCaller(intervals, test, tver, opts):\n",
    "\n",
    "    # skip?\n",
    "    if len(OP_TVER_ONLY) and OP_TVER_ONLY != tver:\n",
    "        return None\n",
    "    \n",
    "    # prepare\n",
    "    outFolder = \"%s/%s/%s\" % (OUT, test, tver)\n",
    "    tpref = \"%s/%s-%s\" % (outFolder, test, tver)\n",
    "    vcf = \"%s.vcf\" % tpref\n",
    "    bam = \"%s.bam\" % tpref\n",
    "    os.system(\"mkdir -p \" + outFolder)\n",
    "\n",
    "    # build main command\n",
    "    cmd = \"\"\"\n",
    "            %s HaplotypeCaller -R %s -I %s -intervals %s \\\n",
    "            --likelihood-calculation-engine FlowBased \\\n",
    "            --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false \\\n",
    "            -O %s \\\n",
    "            --bam-output %s \\\n",
    "            %s \\\n",
    "    \"\"\" % (GATK, REF, READS, intervals, vcf, bam, opts)\n",
    "\n",
    "    # execute it\n",
    "    print(\"Running %s-%s over %s opts %s\" % (test, tver, intervals, opts))\n",
    "    lines = []\n",
    "    if OP_PRINT_CMD:\n",
    "        print(cmd)\n",
    "    with subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True) as proc:\n",
    "        for line in io.TextIOWrapper(proc.stdout, encoding=\"utf-8\"):\n",
    "            lines.append(line)\n",
    "            if \"ProgressMeter\" in line:\n",
    "                print(line, end = '')\n",
    "\n",
    "    # analyze vcf\n",
    "    anal_cmd = \"bcftools stats %s  | grep -E '^SN' | cut -f 3-4 | sed 's/number of //g' | sed 's/ /-/g' | sed 's/://g'\" \\\n",
    "                    % vcf\n",
    "    anal = {\"_vcf\": vcf, \"_test\": test, \"_tver\": tver, \"_intervals\": intervals, \"_opts\": opts}\n",
    "    with subprocess.Popen(anal_cmd, stdout=subprocess.PIPE, shell=True) as proc:\n",
    "        for line in io.TextIOWrapper(proc.stdout, encoding=\"utf-8\"):\n",
    "            toks = line.split()\n",
    "            anal[toks[0].lower()] = toks[1]\n",
    "    pprint.pprint(anal)\n",
    "    with io.open(tpref + \"-anal.json\", 'w') as file:\n",
    "        file.write(json.dumps(anal))\n",
    "\n",
    "    return anal\n",
    "\n",
    "# perform a test\n",
    "if OP_ANAL:\n",
    "    HaplotypeCaller(INTERVALS, TESTNAME, \"baseline\", \"\")\n",
    "    HaplotypeCaller(INTERVALS, TESTNAME, \"collapse12m\", \"--ultima-assembly-collapse-hmer-size 12 --ultima-flow-matrix-mods 10,12,11,12\")\n",
    "    HaplotypeCaller(INTERVALS, TESTNAME, \"collapse10m\", \"--ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods 11,10,12,10\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
