{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running test1-collapse10 over chr20:6000000-7000000 opts --ultima-assembly-collapse-hmer-size 10\n",
      "12:34:25.417 INFO  ProgressMeter - Starting traversal\n",
      "12:34:25.418 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "12:34:35.946 INFO  ProgressMeter -        chr20:6018361              0.2                   140            797.9\n",
      "12:34:46.184 INFO  ProgressMeter -        chr20:6042041              0.3                   320            924.6\n",
      "12:34:56.688 INFO  ProgressMeter -        chr20:6068737              0.5                   520            997.8\n",
      "12:35:06.956 INFO  ProgressMeter -        chr20:6096849              0.7                   720           1040.0\n",
      "12:35:17.037 INFO  ProgressMeter -        chr20:6126752              0.9                   930           1081.0\n",
      "12:35:27.187 INFO  ProgressMeter -        chr20:6147994              1.0                  1100           1068.5\n",
      "12:35:37.410 INFO  ProgressMeter -        chr20:6173939              1.2                  1290           1075.1\n",
      "12:35:47.444 INFO  ProgressMeter -        chr20:6205456              1.4                  1510           1104.5\n",
      "12:35:57.599 INFO  ProgressMeter -        chr20:6233800              1.5                  1710           1113.0\n",
      "12:36:08.254 INFO  ProgressMeter -        chr20:6259022              1.7                  1900           1108.6\n",
      "12:36:18.893 INFO  ProgressMeter -        chr20:6279116              1.9                  2050           1083.9\n",
      "12:36:29.126 INFO  ProgressMeter -        chr20:6302636              2.1                  2230           1081.6\n",
      "12:36:39.148 INFO  ProgressMeter -        chr20:6336989              2.2                  2460           1103.7\n",
      "12:36:49.771 INFO  ProgressMeter -        chr20:6369139              2.4                  2700           1122.2\n",
      "12:37:00.335 INFO  ProgressMeter -        chr20:6395162              2.6                  2910           1127.1\n",
      "12:37:10.970 INFO  ProgressMeter -        chr20:6416131              2.8                  3080           1116.3\n",
      "12:37:20.989 INFO  ProgressMeter -        chr20:6437579              2.9                  3250           1110.7\n",
      "12:37:30.995 INFO  ProgressMeter -        chr20:6457004              3.1                  3420           1105.7\n",
      "12:37:41.636 INFO  ProgressMeter -        chr20:6483050              3.3                  3620           1106.9\n",
      "12:37:51.670 INFO  ProgressMeter -        chr20:6502627              3.4                  3780           1099.6\n",
      "12:38:01.775 INFO  ProgressMeter -        chr20:6527731              3.6                  3970           1101.0\n",
      "12:38:11.833 INFO  ProgressMeter -        chr20:6558847              3.8                  4190           1110.4\n",
      "12:38:22.219 INFO  ProgressMeter -        chr20:6581508              3.9                  4360           1104.7\n",
      "12:38:32.595 INFO  ProgressMeter -        chr20:6608268              4.1                  4550           1104.5\n",
      "12:38:43.026 INFO  ProgressMeter -        chr20:6638826              4.3                  4770           1111.0\n",
      "12:38:53.684 INFO  ProgressMeter -        chr20:6664216              4.5                  4940           1104.9\n",
      "12:39:04.007 INFO  ProgressMeter -        chr20:6694157              4.6                  5170           1113.5\n",
      "12:39:14.180 INFO  ProgressMeter -        chr20:6715096              4.8                  5340           1109.6\n",
      "12:39:24.277 INFO  ProgressMeter -        chr20:6740574              5.0                  5530           1110.2\n",
      "12:39:34.615 INFO  ProgressMeter -        chr20:6767316              5.2                  5730           1111.9\n",
      "12:39:45.297 INFO  ProgressMeter -        chr20:6788198              5.3                  5900           1106.7\n",
      "12:39:55.697 INFO  ProgressMeter -        chr20:6813144              5.5                  6090           1106.3\n",
      "12:40:05.908 INFO  ProgressMeter -        chr20:6846906              5.7                  6320           1113.7\n",
      "12:40:16.279 INFO  ProgressMeter -        chr20:6873421              5.8                  6530           1116.7\n",
      "12:40:26.399 INFO  ProgressMeter -        chr20:6898891              6.0                  6710           1115.3\n",
      "12:40:37.040 INFO  ProgressMeter -        chr20:6928507              6.2                  6940           1120.5\n",
      "12:40:47.438 INFO  ProgressMeter -        chr20:6948573              6.4                  7100           1115.1\n",
      "12:40:57.725 INFO  ProgressMeter -        chr20:6976277              6.5                  7310           1118.0\n",
      "12:41:06.555 INFO  ProgressMeter -        chr20:6999805              6.7                  7491           1120.5\n",
      "12:41:06.555 INFO  ProgressMeter - Traversal complete. Processed 7491 total regions in 6.7 minutes.\n",
      "{'_intervals': 'chr20:6000000-7000000',\n",
      " '_opts': '--ultima-assembly-collapse-hmer-size 10',\n",
      " '_test': 'test1',\n",
      " '_tver': 'collapse10',\n",
      " '_vcf': 'output/test1/collapse10/test1-collapse10.vcf',\n",
      " 'indels': '1650',\n",
      " 'mnps': '0',\n",
      " 'multiallelic-sites': '162',\n",
      " 'multiallelic-snp-sites': '1',\n",
      " 'no-alts': '0',\n",
      " 'others': '0',\n",
      " 'records': '3071',\n",
      " 'samples': '1',\n",
      " 'snps': '1423'}\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import io\n",
    "import pprint\n",
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "# operations\n",
    "OP_ANAL=True\n",
    "OP_PRINT_CMD=False\n",
    "OP_TVER_ONLY=\"collapse10\"\n",
    "\n",
    "# globals\n",
    "GATK=\"../../gatk\"\n",
    "OUT=\"output\"\n",
    "READS=\"data/chr20.flow.bam\"\n",
    "REF=\"../ref/Homo_sapiens_assembly38.fasta\"\n",
    "INTERVALS=\"chr20:6000000-7000000\"\n",
    "\n",
    "# functions\n",
    "def HaplotypeCaller(intervals, test, tver, opts):\n",
    "\n",
    "    # skip?\n",
    "    if len(OP_TVER_ONLY) and OP_TVER_ONLY != tver:\n",
    "        return None\n",
    "    \n",
    "    # prepare\n",
    "    outFolder = \"%s/%s/%s\" % (OUT, test, tver)\n",
    "    tpref = \"%s/%s-%s\" % (outFolder, test, tver)\n",
    "    vcf = \"%s.vcf\" % tpref\n",
    "    bam = \"%s.bam\" % tpref\n",
    "    os.system(\"mkdir -p \" + outFolder)\n",
    "\n",
    "    # build main command\n",
    "    cmd = \"\"\"\n",
    "            %s HaplotypeCaller -R %s -I %s -intervals %s \\\n",
    "            --likelihood-calculation-engine FlowBased \\\n",
    "            --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false \\\n",
    "            -O %s \\\n",
    "            --bam-output %s \\\n",
    "            %s \\\n",
    "    \"\"\" % (GATK, REF, READS, intervals, vcf, bam, opts)\n",
    "\n",
    "    # execute it\n",
    "    print(\"Running %s-%s over %s opts %s\" % (test, tver, intervals, opts))\n",
    "    lines = []\n",
    "    if OP_PRINT_CMD:\n",
    "        print(cmd)\n",
    "    with subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True) as proc:\n",
    "        for line in io.TextIOWrapper(proc.stdout, encoding=\"utf-8\"):\n",
    "            lines.append(line)\n",
    "            if \"ProgressMeter\" in line:\n",
    "                print(line, end = '')\n",
    "\n",
    "    # analyze vcf\n",
    "    anal_cmd = \"bcftools stats %s  | grep -E '^SN' | cut -f 3-4 | sed 's/number of //g' | sed 's/ /-/g' | sed 's/://g'\" \\\n",
    "                    % vcf\n",
    "    anal = {\"_vcf\": vcf, \"_test\": test, \"_tver\": tver, \"_intervals\": intervals, \"_opts\": opts}\n",
    "    with subprocess.Popen(anal_cmd, stdout=subprocess.PIPE, shell=True) as proc:\n",
    "        for line in io.TextIOWrapper(proc.stdout, encoding=\"utf-8\"):\n",
    "            toks = line.split()\n",
    "            anal[toks[0].lower()] = toks[1]\n",
    "    pprint.pprint(anal)\n",
    "    with io.open(tpref + \"-anal.json\", 'w') as file:\n",
    "        file.write(json.dumps(anal))\n",
    "\n",
    "    return anal\n",
    "\n",
    "# perform a test\n",
    "if OP_ANAL:\n",
    "    HaplotypeCaller(INTERVALS, \"test1\", \"baseline\", \"\")\n",
    "    HaplotypeCaller(INTERVALS, \"test1\", \"baseline12m\", \"--ultima-flow-matrix-mods 10,12,11,12\")\n",
    "    HaplotypeCaller(INTERVALS, \"test1\", \"baseline10m\", \"--ultima-flow-matrix-mods 8,10,9,10,11,10,12,10\")\n",
    "\n",
    "    HaplotypeCaller(INTERVALS, \"test1\", \"collapse12\", \"--ultima-assembly-collapse-hmer-size 12\")\n",
    "    HaplotypeCaller(INTERVALS, \"test1\", \"collapse12m\", \"--ultima-assembly-collapse-hmer-size 12 --ultima-flow-matrix-mods 10,12,11,12\")\n",
    "    \n",
    "    HaplotypeCaller(INTERVALS, \"test1\", \"collapse10\", \"--ultima-assembly-collapse-hmer-size 10\")\n",
    "    HaplotypeCaller(INTERVALS, \"test1\", \"collapse10m\", \"--ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods 8,10,9,10,11,10,12,10\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
