{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running test3-collapse10m over chr9:6600000-7000000 opts --ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods 11,10,12,10\n",
      "\n",
      "            ../../gatk HaplotypeCaller -R ../ref/Homo_sapiens_assembly38.fasta -I data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam -intervals chr9:6600000-7000000             --likelihood-calculation-engine FlowBased             --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false -mbq 0             -O output/test3/collapse10m/test3-collapse10m.vcf             --bam-output output/test3/collapse10m/test3-collapse10m.bam             --ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods 11,10,12,10     \n",
      "06:31:30.797 INFO  ProgressMeter - Starting traversal\n",
      "06:31:30.797 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "06:31:41.139 INFO  ProgressMeter -         chr9:6613830              0.2                   130            754.2\n",
      "06:31:51.166 INFO  ProgressMeter -         chr9:6643795              0.3                   360           1060.4\n",
      "06:32:01.343 INFO  ProgressMeter -         chr9:6664080              0.5                   530           1041.1\n",
      "06:32:11.572 INFO  ProgressMeter -         chr9:6684765              0.7                   700           1030.0\n",
      "06:32:21.688 INFO  ProgressMeter -         chr9:6707315              0.8                   880           1037.5\n",
      "06:32:32.035 INFO  ProgressMeter -         chr9:6722866              1.0                  1000            979.8\n",
      "06:32:42.122 INFO  ProgressMeter -         chr9:6748413              1.2                  1210           1017.9\n",
      "06:32:53.032 INFO  ProgressMeter -         chr9:6779198              1.4                  1460           1065.2\n",
      "06:33:03.133 INFO  ProgressMeter -         chr9:6796612              1.5                  1600           1039.7\n",
      "06:33:14.402 INFO  ProgressMeter -         chr9:6832494              1.7                  1860           1077.2\n",
      "06:33:24.612 INFO  ProgressMeter -         chr9:6857104              1.9                  2040           1075.4\n",
      "06:33:34.813 INFO  ProgressMeter -         chr9:6895715              2.1                  2300           1112.8\n",
      "06:33:44.874 INFO  ProgressMeter -         chr9:6924726              2.2                  2500           1118.8\n",
      "06:33:55.076 INFO  ProgressMeter -         chr9:6940128              2.4                  2630           1093.7\n",
      "06:34:05.114 INFO  ProgressMeter -         chr9:6963507              2.6                  2810           1092.6\n",
      "06:34:15.546 INFO  ProgressMeter -         chr9:6991063              2.7                  3040           1107.1\n",
      "06:34:21.732 INFO  ProgressMeter -         chr9:6999114              2.8                  3117           1094.1\n",
      "06:34:21.732 INFO  ProgressMeter - Traversal complete. Processed 3117 total regions in 2.8 minutes.\n",
      "{'_intervals': 'chr9:6600000-7000000',\n",
      " '_opts': '--ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods '\n",
      "          '11,10,12,10',\n",
      " '_test': 'test3',\n",
      " '_tver': 'collapse10m',\n",
      " '_vcf': 'output/test3/collapse10m/test3-collapse10m.vcf',\n",
      " 'indels': '203',\n",
      " 'mnps': '0',\n",
      " 'multiallelic-sites': '2',\n",
      " 'multiallelic-snp-sites': '1',\n",
      " 'no-alts': '0',\n",
      " 'others': '0',\n",
      " 'records': '1106',\n",
      " 'samples': '1',\n",
      " 'snps': '903'}\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import io\n",
    "import pprint\n",
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "# operations\n",
    "OP_ANAL=True\n",
    "OP_PRINT_CMD=True\n",
    "OP_TVER_ONLY=\"\"\n",
    "\n",
    "# globals\n",
    "GATK=\"../../gatk\"\n",
    "OUT=\"output\"\n",
    "REF=\"../ref/Homo_sapiens_assembly38.fasta\"\n",
    "\n",
    "#READS=\"data/chr20.flow.bam\"\n",
    "#INTERVALS=\"chr20:6000000-7000000\"\n",
    "#TESTNAME=\"test1\"\n",
    "\n",
    "#READS=\"data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam\"\n",
    "#INTERVALS=\"chr9:1700000-2000000\"\n",
    "#TESTNAME=\"test2\"\n",
    "\n",
    "READS=\"data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam\"\n",
    "INTERVALS=\"chr9:6600000-7000000\"\n",
    "TESTNAME=\"test3\"\n",
    "\n",
    "#INTERVALS=\"chr9:6943300-6943900 -L chr9:6745098-6746098 -L chr9:6688000-6689000\"\n",
    "\n",
    "# functions\n",
    "def HaplotypeCaller(intervals, test, tver, opts):\n",
    "\n",
    "    # skip?\n",
    "    if len(OP_TVER_ONLY) and OP_TVER_ONLY != tver:\n",
    "        return None\n",
    "    \n",
    "    # prepare\n",
    "    outFolder = \"%s/%s/%s\" % (OUT, test, tver)\n",
    "    tpref = \"%s/%s-%s\" % (outFolder, test, tver)\n",
    "    vcf = \"%s.vcf\" % tpref\n",
    "    bam = \"%s.bam\" % tpref\n",
    "    os.system(\"mkdir -p \" + outFolder)\n",
    "\n",
    "    # build main command\n",
    "    cmd = \"\"\"\n",
    "            %s HaplotypeCaller -R %s -I %s -intervals %s \\\n",
    "            --likelihood-calculation-engine FlowBased \\\n",
    "            --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false -mbq 0 \\\n",
    "            -O %s \\\n",
    "            --bam-output %s \\\n",
    "            %s \\\n",
    "    \"\"\" % (GATK, REF, READS, intervals, vcf, bam, opts)\n",
    "\n",
    "    # execute it\n",
    "    print(\"Running %s-%s over %s opts %s\" % (test, tver, intervals, opts))\n",
    "    lines = []\n",
    "    if OP_PRINT_CMD:\n",
    "        print(cmd)\n",
    "    with subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True) as proc:\n",
    "        for line in io.TextIOWrapper(proc.stdout, encoding=\"utf-8\"):\n",
    "            lines.append(line)\n",
    "            if \"ProgressMeter\" in line:\n",
    "                print(line, end = '')\n",
    "\n",
    "    # analyze vcf\n",
    "    anal_cmd = \"bcftools stats %s  | grep -E '^SN' | cut -f 3-4 | sed 's/number of //g' | sed 's/ /-/g' | sed 's/://g'\" \\\n",
    "                    % vcf\n",
    "    anal = {\"_vcf\": vcf, \"_test\": test, \"_tver\": tver, \"_intervals\": intervals, \"_opts\": opts}\n",
    "    with subprocess.Popen(anal_cmd, stdout=subprocess.PIPE, shell=True) as proc:\n",
    "        for line in io.TextIOWrapper(proc.stdout, encoding=\"utf-8\"):\n",
    "            toks = line.split()\n",
    "            anal[toks[0].lower()] = toks[1]\n",
    "    pprint.pprint(anal)\n",
    "    with io.open(tpref + \"-anal.json\", 'w') as file:\n",
    "        file.write(json.dumps(anal))\n",
    "\n",
    "    return anal\n",
    "\n",
    "# perform a test\n",
    "if OP_ANAL:\n",
    "    #HaplotypeCaller(INTERVALS, TESTNAME, \"baseline\", \"\")\n",
    "    #HaplotypeCaller(INTERVALS, TESTNAME, \"collapse12m\", \"--ultima-assembly-collapse-hmer-size 12 --ultima-flow-matrix-mods 10,12,11,12\")\n",
    "    HaplotypeCaller(INTERVALS, TESTNAME, \"collapse10m\", \"--ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods 11,10,12,10\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
