{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running test2-baseline over chr9:1700000-2000000 opts \n",
      "\n",
      "            ../../gatk HaplotypeCaller -R ../ref/Homo_sapiens_assembly38.fasta -I data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam -intervals chr9:1700000-2000000             --likelihood-calculation-engine FlowBased             --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false             -O output/test2/baseline/test2-baseline.vcf             --bam-output output/test2/baseline/test2-baseline.bam                  \n",
      "09:22:11.333 INFO  ProgressMeter - Starting traversal\n",
      "09:22:11.333 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "09:22:21.383 INFO  ProgressMeter -         chr9:1736048              0.2                   290           1731.3\n",
      "09:22:31.457 INFO  ProgressMeter -         chr9:1847992              0.3                  1080           3220.2\n",
      "09:22:41.604 INFO  ProgressMeter -         chr9:1961070              0.5                  1920           3805.6\n",
      "09:22:45.508 INFO  ProgressMeter -         chr9:1998703              0.6                  2209           3878.3\n",
      "09:22:45.508 INFO  ProgressMeter - Traversal complete. Processed 2209 total regions in 0.6 minutes.\n",
      "{'_intervals': 'chr9:1700000-2000000',\n",
      " '_opts': '',\n",
      " '_test': 'test2',\n",
      " '_tver': 'baseline',\n",
      " '_vcf': 'output/test2/baseline/test2-baseline.vcf',\n",
      " 'indels': '124',\n",
      " 'mnps': '0',\n",
      " 'multiallelic-sites': '3',\n",
      " 'multiallelic-snp-sites': '1',\n",
      " 'no-alts': '0',\n",
      " 'others': '0',\n",
      " 'records': '728',\n",
      " 'samples': '1',\n",
      " 'snps': '604'}\n",
      "Running test2-collapse12m over chr9:1700000-2000000 opts --ultima-assembly-collapse-hmer-size 12 --ultima-flow-matrix-mods 10,12,11,12\n",
      "\n",
      "            ../../gatk HaplotypeCaller -R ../ref/Homo_sapiens_assembly38.fasta -I data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam -intervals chr9:1700000-2000000             --likelihood-calculation-engine FlowBased             --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false             -O output/test2/collapse12m/test2-collapse12m.vcf             --bam-output output/test2/collapse12m/test2-collapse12m.bam             --ultima-assembly-collapse-hmer-size 12 --ultima-flow-matrix-mods 10,12,11,12     \n",
      "09:22:50.292 INFO  ProgressMeter - Starting traversal\n",
      "09:22:50.293 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "09:23:00.609 INFO  ProgressMeter -         chr9:1734515              0.2                   280           1628.5\n",
      "09:23:10.692 INFO  ProgressMeter -         chr9:1860000              0.3                  1160           3411.9\n",
      "09:23:20.812 INFO  ProgressMeter -         chr9:1968084              0.5                  1980           3892.7\n",
      "09:23:23.623 INFO  ProgressMeter -         chr9:1998703              0.6                  2209           3976.6\n",
      "09:23:23.623 INFO  ProgressMeter - Traversal complete. Processed 2209 total regions in 0.6 minutes.\n",
      "{'_intervals': 'chr9:1700000-2000000',\n",
      " '_opts': '--ultima-assembly-collapse-hmer-size 12 --ultima-flow-matrix-mods '\n",
      "          '10,12,11,12',\n",
      " '_test': 'test2',\n",
      " '_tver': 'collapse12m',\n",
      " '_vcf': 'output/test2/collapse12m/test2-collapse12m.vcf',\n",
      " 'indels': '117',\n",
      " 'mnps': '0',\n",
      " 'multiallelic-sites': '2',\n",
      " 'multiallelic-snp-sites': '1',\n",
      " 'no-alts': '0',\n",
      " 'others': '0',\n",
      " 'records': '718',\n",
      " 'samples': '1',\n",
      " 'snps': '601'}\n",
      "Running test2-collapse10m over chr9:1700000-2000000 opts --ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods 11,10,12,10\n",
      "\n",
      "            ../../gatk HaplotypeCaller -R ../ref/Homo_sapiens_assembly38.fasta -I data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam -intervals chr9:1700000-2000000             --likelihood-calculation-engine FlowBased             --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false             -O output/test2/collapse10m/test2-collapse10m.vcf             --bam-output output/test2/collapse10m/test2-collapse10m.bam             --ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods 11,10,12,10     \n",
      "09:23:28.294 INFO  ProgressMeter - Starting traversal\n",
      "09:23:28.294 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute\n",
      "09:23:38.324 INFO  ProgressMeter -         chr9:1734515              0.2                   280           1675.0\n",
      "09:23:48.369 INFO  ProgressMeter -         chr9:1863959              0.3                  1190           3556.7\n",
      "09:23:58.378 INFO  ProgressMeter -         chr9:1968084              0.5                  1980           3948.9\n",
      "09:24:01.167 INFO  ProgressMeter -         chr9:1998703              0.5                  2209           4031.9\n",
      "09:24:01.167 INFO  ProgressMeter - Traversal complete. Processed 2209 total regions in 0.5 minutes.\n",
      "{'_intervals': 'chr9:1700000-2000000',\n",
      " '_opts': '--ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods '\n",
      "          '11,10,12,10',\n",
      " '_test': 'test2',\n",
      " '_tver': 'collapse10m',\n",
      " '_vcf': 'output/test2/collapse10m/test2-collapse10m.vcf',\n",
      " 'indels': '118',\n",
      " 'mnps': '0',\n",
      " 'multiallelic-sites': '2',\n",
      " 'multiallelic-snp-sites': '1',\n",
      " 'no-alts': '0',\n",
      " 'others': '0',\n",
      " 'records': '716',\n",
      " 'samples': '1',\n",
      " 'snps': '598'}\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import io\n",
    "import pprint\n",
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "# operations\n",
    "OP_ANAL=True\n",
    "OP_PRINT_CMD=True\n",
    "OP_TVER_ONLY=\"\"\n",
    "\n",
    "# globals\n",
    "GATK=\"../../gatk\"\n",
    "OUT=\"output\"\n",
    "REF=\"../ref/Homo_sapiens_assembly38.fasta\"\n",
    "\n",
    "#READS=\"data/chr20.flow.bam\"\n",
    "#INTERVALS=\"chr20:6000000-7000000\"\n",
    "\n",
    "READS=\"data/200603/140258_probability_E5_Z7.aligned.duplicate_marked.sorted.bam\"\n",
    "INTERVALS=\"chr9:1700000-2000000\"\n",
    "\n",
    "# functions\n",
    "def HaplotypeCaller(intervals, test, tver, opts):\n",
    "\n",
    "    # skip?\n",
    "    if len(OP_TVER_ONLY) and OP_TVER_ONLY != tver:\n",
    "        return None\n",
    "    \n",
    "    # prepare\n",
    "    outFolder = \"%s/%s/%s\" % (OUT, test, tver)\n",
    "    tpref = \"%s/%s-%s\" % (outFolder, test, tver)\n",
    "    vcf = \"%s.vcf\" % tpref\n",
    "    bam = \"%s.bam\" % tpref\n",
    "    os.system(\"mkdir -p \" + outFolder)\n",
    "\n",
    "    # build main command\n",
    "    cmd = \"\"\"\n",
    "            %s HaplotypeCaller -R %s -I %s -intervals %s \\\n",
    "            --likelihood-calculation-engine FlowBased \\\n",
    "            --bam-writer-type ALL_POSSIBLE_HAPLOTYPES --debug-assembly false \\\n",
    "            -O %s \\\n",
    "            --bam-output %s \\\n",
    "            %s \\\n",
    "    \"\"\" % (GATK, REF, READS, intervals, vcf, bam, opts)\n",
    "\n",
    "    # execute it\n",
    "    print(\"Running %s-%s over %s opts %s\" % (test, tver, intervals, opts))\n",
    "    lines = []\n",
    "    if OP_PRINT_CMD:\n",
    "        print(cmd)\n",
    "    with subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True) as proc:\n",
    "        for line in io.TextIOWrapper(proc.stdout, encoding=\"utf-8\"):\n",
    "            lines.append(line)\n",
    "            if \"ProgressMeter\" in line:\n",
    "                print(line, end = '')\n",
    "\n",
    "    # analyze vcf\n",
    "    anal_cmd = \"bcftools stats %s  | grep -E '^SN' | cut -f 3-4 | sed 's/number of //g' | sed 's/ /-/g' | sed 's/://g'\" \\\n",
    "                    % vcf\n",
    "    anal = {\"_vcf\": vcf, \"_test\": test, \"_tver\": tver, \"_intervals\": intervals, \"_opts\": opts}\n",
    "    with subprocess.Popen(anal_cmd, stdout=subprocess.PIPE, shell=True) as proc:\n",
    "        for line in io.TextIOWrapper(proc.stdout, encoding=\"utf-8\"):\n",
    "            toks = line.split()\n",
    "            anal[toks[0].lower()] = toks[1]\n",
    "    pprint.pprint(anal)\n",
    "    with io.open(tpref + \"-anal.json\", 'w') as file:\n",
    "        file.write(json.dumps(anal))\n",
    "\n",
    "    return anal\n",
    "\n",
    "# perform a test\n",
    "if OP_ANAL:\n",
    "    HaplotypeCaller(INTERVALS, \"test2\", \"baseline\", \"\")\n",
    "    HaplotypeCaller(INTERVALS, \"test2\", \"collapse12m\", \"--ultima-assembly-collapse-hmer-size 12 --ultima-flow-matrix-mods 10,12,11,12\")\n",
    "    HaplotypeCaller(INTERVALS, \"test2\", \"collapse10m\", \"--ultima-assembly-collapse-hmer-size 10 --ultima-flow-matrix-mods 11,10,12,10\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
